timer/api/timer.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/parser/duration"
+)
+
+// SchedPolicyType is the type of the event schedule policy
+type SchedPolicyType string
+
+const (
+	// SchedEventInterval indicates to schedule events every fixed interval
+	SchedEventInterval SchedPolicyType = "INTERVAL"
+)
+
+// SchedEventPolicy is an interface to tell the runtime how to schedule a timer's events
+type SchedEventPolicy interface {
+	// NextEventTime returns the time to schedule the next timer event. If the second return value is true,
+	// it means we have a next event schedule after `watermark`. Otherwise, it means there is no more event after `watermark`.
+	NextEventTime(watermark time.Time) (time.Time, bool)
+}
+
+// SchedIntervalPolicy implements SchedEventPolicy, it is the policy of type `SchedEventInterval`
+type SchedIntervalPolicy struct {
```

Comment:
I wonder what's the expected behavior if the client found it's too late to start the job (e.g. if the cluster is down for one or two days.). If they are only expected to execute when the time is within a small offset around the "expected" schedule (e.g. `*/5 * * * *` shouldn't happen at 00:02), a function like "NearestSchedTime" (or other better names) which will return the nearest "expected" schedule time would be really helpful.

The only difference between `NearestSchedTime` and for-loop with `NextEventTime` is that the  `NearestSchedTime` can calculate much faster if the timer is regular (e.g. for every 5 seconds). If we don't have this, we'll need to introduce a new range (or limitation) to avoid the "for-loop next" calculates forever, which will also make the expected behavior hard to describe/learn (See the [`.spec.startingDeadlineSeconds`](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#job-creation) in kubernetes cronjob as a counter example :man_facepalming:  )

Reply:
Good catch! I think there is a principle in the framework that there is only one event triggering at the same time. For example, if the timer is set to trigger at the first minute of each hour but the cluster is down for one or two days. When the cluster recovered, there was still only one event triggered even if we have many events after the watermark. 

Whether to execute all these jobs or ignore some "expired" depends on the user:

1. If the user wants to execute all passed jobs, he should compute `NextEventTime` again and again until the returned event time after now.
2. If the user wants to ignore some jobs after a certain time, he can pass that time instead of the watermark to `NextEventTime` to compute the next event time after it.

When processing one event, a user can submit one or more jobs concurrency, it dependens to him.